onnxruntime/core/framework/execution_frame.cc

```diff
@@ -750,6 +700,33 @@ Status ExecutionFrame::CreateNodeOutputMLValueImpl(OrtValue& ort_value, int ort_
   return AllocateAsPerAllocationPlan(ort_value, ort_value_idx, shape, nnz);
 }
 
+void ExecutionFrame::VerifyOutputSizes(int output_index, const Node& node, const TensorShape& output_shape) {
+  const NodeArg* output_def = node.OutputDefs()[output_index];
+  const auto* expected_shape = output_def->Shape();
+  if (expected_shape == nullptr) {
+    // model didn't specify shape and shape inferencing wasn't able to calculate it so nothing to compare against
+    return;
+  }
+
+  size_t expected_rank = expected_shape->dim_size();
```

Comment:
nit: can be const size_t expected_rank and the 'end' variable in the loop below is not required.